File: cobble-binding/cobble-cpp/cpp/multi_get.cpp
#include <cobble/database.hpp>

#include <cstddef>
#include <cstdint>
#include <limits>
#include <type_traits>
#include <utility>
#include <vector>

#include "detail/convert.hpp"
#include "detail/error.hpp"
#include "detail/impl.hpp"
#include "detail/options.hpp"

namespace cobble {
namespace {

// This private descriptor is borrowed by Rust only for the synchronous bridge
// call. Payload bytes are never concatenated or copied.
struct KeyDescriptor {
  std::uint16_t bucket;
  std::uint16_t padding;
  const Byte* data;
  std::size_t length;
};

static_assert(std::is_standard_layout_v<KeyDescriptor>);
constexpr std::size_t AlignUp(std::size_t value, std::size_t alignment) {
  return (value + alignment - 1) / alignment * alignment;
}
constexpr std::size_t kDescriptorAlignment =
    alignof(const Byte*) > alignof(std::size_t) ? alignof(const Byte*)
                                                : alignof(std::size_t);
constexpr std::size_t kDataOffset =
    AlignUp(2 * sizeof(std::uint16_t), alignof(const Byte*));
constexpr std::size_t kLengthOffset =
    AlignUp(kDataOffset + sizeof(const Byte*), alignof(std::size_t));
constexpr std::size_t kDescriptorSize =
    AlignUp(kLengthOffset + sizeof(std::size_t), kDescriptorAlignment);
static_assert(offsetof(KeyDescriptor, bucket) == 0);
static_assert(offsetof(KeyDescriptor, padding) == sizeof(std::uint16_t));
static_assert(offsetof(KeyDescriptor, data) == kDataOffset);
static_assert(offsetof(KeyDescriptor, length) == kLengthOffset);
static_assert(alignof(KeyDescriptor) == kDescriptorAlignment);
static_assert(sizeof(KeyDescriptor) == kDescriptorSize);

template <typename Call>
rust::Box<ffi::NativeMultiGetResult> CallMultiGet(
    std::span<const MultiGetKey> keys, const ReadOptions& options, Call&& call) {
  if (keys.size() > std::numeric_limits<std::uint64_t>::max()) {
    throw Error(ErrorCode::kInput, "multi-get key count does not fit in u64");
  }

  std::vector<KeyDescriptor> descriptors;
  descriptors.reserve(keys.size());
  for (const auto& key : keys) {
    descriptors.push_back({key.bucket, 0, key.key.data(), key.key.size()});
  }
  const auto native_options = detail::ToNative(options);
  return detail::Translate([&] {
    return call(reinterpret_cast<std::size_t>(descriptors.data()),
                descriptors.size(), native_options);
  });
}

}  // namespace

struct OwnedMultiGetResult::Impl {
  explicit Impl(rust::Box<ffi::NativeMultiGetResult> native_result)
      : native(std::move(native_result)) {}

  rust::Box<ffi::NativeMultiGetResult> native;
};

OwnedMultiGetResult::OwnedMultiGetResult(std::unique_ptr<Impl> impl) noexcept
    : impl_(std::move(impl)) {}
OwnedMultiGetResult::OwnedMultiGetResult(OwnedMultiGetResult&&) noexcept =
    default;
OwnedMultiGetResult& OwnedMultiGetResult::operator=(
    OwnedMultiGetResult&&) noexcept = default;
OwnedMultiGetResult::~OwnedMultiGetResult() = default;

std::size_t OwnedMultiGetResult::row_count() const noexcept {
  if (!impl_) {
    return 0;
  }
  const auto count = ffi::native_multi_get_row_count(*impl_->native);
  return count > std::numeric_limits<std::size_t>::max()
             ? std::numeric_limits<std::size_t>::max()
             : static_cast<std::size_t>(count);
}

bool OwnedMultiGetResult::found(std::size_t row) const {
  return impl_ && ffi::native_multi_get_found(*impl_->native, row);
}

std::size_t OwnedMultiGetResult::column_count(std::size_t row) const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState,
                "OwnedMultiGetResult has been moved from");
  }
  return detail::ToSize(detail::Translate([&] {
                          return ffi::native_multi_get_column_count(
                              *impl_->native, row);
                        }),
                        "multi-get column count");
}

bool OwnedMultiGetResult::has_column(std::size_t row,
                                     std::size_t column) const {
  return impl_ &&
         ffi::native_multi_get_has_column(*impl_->native, row, column);
}

BytesView OwnedMultiGetResult::column(std::size_t row,
                                      std::size_t column) const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState,
                "OwnedMultiGetResult has been moved from");
  }
  return detail::Translate([&] {
    return detail::ToView(
        ffi::native_multi_get_column(*impl_->native, row, column));
  });
}

OwnedMultiGetResult Database::MultiGet(
    std::span<const MultiGetKey> keys, const ReadOptions& options) const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Database has been moved from");
  }
  auto native = CallMultiGet(keys, options, [&](auto address, auto count,
                                                const auto& native_options) {
    return ffi::native_database_multi_get(
        *impl_->native, address, count, native_options);
  });
  return OwnedMultiGetResult(
      std::make_unique<OwnedMultiGetResult::Impl>(std::move(native)));
}

OwnedMultiGetResult Db::MultiGet(std::span<const MultiGetKey> keys,
                                 const ReadOptions& options) const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Db has been moved from");
  }
  auto native = CallMultiGet(keys, options, [&](auto address, auto count,
                                                const auto& native_options) {
    return ffi::native_sharded_database_multi_get(
        *impl_->native, address, count, native_options);
  });
  return OwnedMultiGetResult(
      std::make_unique<OwnedMultiGetResult::Impl>(std::move(native)));
}

OwnedMultiGetResult ReadOnlyDb::MultiGet(
    std::span<const MultiGetKey> keys, const ReadOptions& options) const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "ReadOnlyDb has been moved from");
  }
  auto native = CallMultiGet(keys, options, [&](auto address, auto count,
                                                const auto& native_options) {
    return ffi::native_read_only_database_multi_get(
        *impl_->native, address, count, native_options);
  });
  return OwnedMultiGetResult(
      std::make_unique<OwnedMultiGetResult::Impl>(std::move(native)));
}

OwnedMultiGetResult Reader::MultiGet(std::span<const MultiGetKey> keys,
                                     const ReadOptions& options) {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Reader has been moved from");
  }
  auto native = CallMultiGet(keys, options, [&](auto address, auto count,
                                                const auto& native_options) {
    return ffi::native_reader_multi_get(*impl_->native, address, count,
                                        native_options);
  });
  return OwnedMultiGetResult(
      std::make_unique<OwnedMultiGetResult::Impl>(std::move(native)));
}

}  // namespace cobble

File: cobble-binding/cobble-cpp/src/multi_get.rs
use bytes::Bytes;
use std::{mem, slice};

use crate::{
    BridgeResult,
    database::NativeDatabase,
    error::{format_cobble_error, input_error},
    ffi,
    options::{checked_u64, checked_usize, to_read_options},
};

#[repr(C)]
struct KeyDescriptor {
    bucket: u16,
    _padding: u16,
    data: *const u8,
    len: usize,
}

pub(crate) struct NativeMultiGetResult {
    rows: Vec<Option<Vec<Option<Bytes>>>>,
}

/// Borrows the caller's key buffers for one synchronous `multi_get` call.
///
/// # Safety
///
/// `descriptor_address` must point to `count` naturally aligned `KeyDescriptor`
/// values created by the private C++ wrapper. Every non-empty descriptor must
/// point to a readable key buffer that remains alive until this function
/// returns. No borrowed address escapes this call.
unsafe fn borrowed_keys<'a>(
    descriptor_address: usize,
    count: u64,
) -> BridgeResult<Vec<(u16, &'a [u8])>> {
    let count = checked_usize(count, "multi-get key count")?;
    if count == 0 {
        return Ok(Vec::new());
    }
    if descriptor_address == 0 {
        return Err(input_error("multi-get descriptors must not be null"));
    }
    if !descriptor_address.is_multiple_of(mem::align_of::<KeyDescriptor>()) {
        return Err(input_error("multi-get descriptors are not aligned"));
    }
    let byte_len = count
        .checked_mul(mem::size_of::<KeyDescriptor>())
        .ok_or_else(|| input_error("multi-get descriptor size overflows"))?;
    if byte_len > isize::MAX as usize {
        return Err(input_error("multi-get descriptor array is too large"));
    }
    descriptor_address
        .checked_add(byte_len)
        .ok_or_else(|| input_error("multi-get descriptor address overflows"))?;

    // SAFETY: the private C++ wrapper satisfies the function contract above;
    // alignment, null, byte-size, and address overflow were checked first.
    let descriptors =
        unsafe { slice::from_raw_parts(descriptor_address as *const KeyDescriptor, count) };
    descriptors
        .iter()
        .map(|entry| {
            if entry.len == 0 {
                return Ok((entry.bucket, &[][..]));
            }
            if entry.data.is_null() {
                return Err(input_error("multi-get key data must not be null"));
            }
            (entry.data as usize)
                .checked_add(entry.len)
                .ok_or_else(|| input_error("multi-get key address overflows"))?;
            // `isize::MAX` is also the maximum length accepted by
            // `slice::from_raw_parts` even when `usize` is wider.
            if entry.len > isize::MAX as usize {
                return Err(input_error("multi-get key length is too large"));
            }
            // SAFETY: the private wrapper keeps each span alive and readable
            // for the duration of this synchronous bridge call.
            Ok((entry.bucket, unsafe {
                slice::from_raw_parts(entry.data, entry.len)
            }))
        })
        .collect()
}

pub(crate) fn native_database_multi_get(
    db: &NativeDatabase,
    descriptor_address: usize,
    count: u64,
    options: &ffi::NativeReadOptions,
) -> BridgeResult<Box<NativeMultiGetResult>> {
    // SAFETY: this function is called only by the private C++ wrapper, and the
    // borrowed descriptors are consumed before returning across the bridge.
    let keys = unsafe { borrowed_keys(descriptor_address, count)? };
    Ok(Box::new(NativeMultiGetResult {
        rows: db
            .db
            .multi_get_with_options(&keys, &to_read_options(options)?)
            .map_err(format_cobble_error)?,
    }))
}

pub(crate) fn native_sharded_database_multi_get(
    db: &crate::sharded_db::NativeShardedDatabase,
    descriptor_address: usize,
    count: u64,
    options: &ffi::NativeReadOptions,
) -> BridgeResult<Box<NativeMultiGetResult>> {
    // SAFETY: this function is called only by the private C++ wrapper, and the
    // borrowed descriptors are consumed before returning across the bridge.
    let keys = unsafe { borrowed_keys(descriptor_address, count)? };
    Ok(Box::new(NativeMultiGetResult {
        rows: db
            .db
            .multi_get_with_options(&keys, &to_read_options(options)?)
            .map_err(format_cobble_error)?,
    }))
}

fn row(rows: &NativeMultiGetResult, index: u64) -> BridgeResult<&Option<Vec<Option<Bytes>>>> {
    rows.rows
        .get(checked_usize(index, "multi-get row index")?)
        .ok_or_else(|| input_error("multi-get row index is out of bounds"))
}

pub(crate) fn native_multi_get_row_count(rows: &NativeMultiGetResult) -> u64 {
    u64::try_from(rows.rows.len()).unwrap_or(u64::MAX)
}

pub(crate) fn native_multi_get_found(rows: &NativeMultiGetResult, index: u64) -> bool {
    row(rows, index).ok().is_some_and(Option::is_some)
}

pub(crate) fn native_multi_get_column_count(
    rows: &NativeMultiGetResult,
    index: u64,
) -> BridgeResult<u64> {
    checked_u64(
        row(rows, index)?.as_ref().map_or(0, Vec::len),
        "multi-get column count",
    )
}

pub(crate) fn native_multi_get_has_column(
    rows: &NativeMultiGetResult,
    index: u64,
    column: u64,
) -> bool {
    checked_usize(column, "column index")
        .ok()
        .and_then(|column| row(rows, index).ok()?.as_ref()?.get(column))
        .is_some_and(Option::is_some)
}

pub(crate) fn native_multi_get_column(
    rows: &NativeMultiGetResult,
    index: u64,
    column: u64,
) -> BridgeResult<&[u8]> {
    let column = checked_usize(column, "column index")?;
    row(rows, index)?
        .as_ref()
        .and_then(|row| row.get(column))
        .and_then(Option::as_ref)
        .map(Bytes::as_ref)
        .ok_or_else(|| input_error("requested multi-get column is absent"))
}
